### ISSUE TITLE
Reduce dependencies number from BlockProcessor

### ISSUE THREAD

**Issue by aboimpinto** [CityOfZion/neo-sharp#350]
1.  IRepository dependency
I don't think is necessary because on this level the persistence level should be abstracted
and some of the dependencies already use it like BlockPersist .. therefore we could use this (for example) and implement the method that is need there
...
2. IBlockPHeaderPersiste and ITransactionPersister
at this level we should not have this dependency but higher level persistence where I just execute: BlockPersister.Persist(block) and it's done ..
all the logic of persist the transactions separate should not be on this level ... why?
image that you are using a relation database, the persistence logic will be different
and we are hooking the persistence RocksDb logic high on our code ... and I don't think this is the right place...
...
3. we are checking if the the block is signed several times .. therefore we will have the OperationManager (new object responsible for the signature outside the DTO)
dependency everywhere ... like here .. inside the persister class .. etc..etc..
here of the block should arrive with hash because it's used to add to the blockPool ...
...
4. ITransactionPool
why this class is responsible to remove something from other pool???
why we have a transaction pool  ??
the transaction aren't added always in the context of a Block?
therefore, when a block is processed, his transaction are processed ...
When we add a block we never add a any transaction to this pool ... why we have to remove ..
this piece of code is a little bit strange ...
the remove call .. should be at least in the same class as the add ... otherwise will will very difficult to follow the code
...
in summary those are the dependencies that I propose to have in this class
IBlockPool
IAsyncDelayer
IBlockPersister             <--- new class that will give us an high level method to persist and retrieve blocks
IBlockOperationsManager        <-- Don't know if this should be here. When the block reach this point should be already been signed (I think)
....
what you think about this possible changes ... do you agree? can we discuss this?

**Comment by osmirnov** [comment]
Well, it is a very thorough analysis. See my answers:
1-2. You suggest introducing an additional layer of the abstraction that will hide implementation details of the Block from BlockProcessor. New BlockPersister will have at least IRepository, IBlockHeaderPersister, and ITransactionPersister in its dependencies. I think it is a good idea to invest in.
3. My opinion is we need to _verify_ signature for blocks we get from outside once and to _assume_ inside that the block has a signature and it was verified before.
4. Transactions are stored on the pool before consensus on a block will kick off. After we have received a block we know which transactions were approved by nodes in consensus and drop only them from the pool if there are any.
